helix-core/src/main/java/org/apache/helix/manager/zk/ZkBucketDataAccessor.java

```diff
@@ -202,126 +218,156 @@ public void disconnect() {
-          } else {
-            System.arraycopy(buckets.get(i), 0, compressedRecord, copyPtr, bucketSize);
-            copyPtr += bucketSize;
-          }
-        }
+    // Compute N - number of buckets
+    int numBuckets = (dataSize + _bucketSize - 1) / _bucketSize;
+    byte[] compressedRecord = new byte[dataSize];
+    String dataPath = path + "/" + versionToRead;
 
-        // Decompress the byte array
-        ByteArrayInputStream byteArrayInputStream = new ByteArrayInputStream(compressedRecord);
-        byte[] serializedRecord;
-        try {
-          serializedRecord = GZipCompressionUtil.uncompress(byteArrayInputStream);
-        } catch (IOException e) {
-          throw new HelixException(String.format("Failed to decompress path: %s!", path), e);
-        }
+    List<String> paths = new ArrayList<>();
+    for (int i = 0; i < numBuckets; i++) {
+      paths.add(dataPath + "/" + i);
+    }
 
-        // Deserialize the record to retrieve the original
-        ZNRecord originalRecord = (ZNRecord) _zkSerializer.deserialize(serializedRecord);
-        return new HelixProperty(originalRecord);
-      } finally {
-        // Critical section for read ends here
-        unlock(path);
+    // Async get
+    List<byte[]> buckets = _zkBaseDataAccessor.get(paths, null, AccessOption.PERSISTENT, true);
+
+    // Combine buckets into one byte array
+    int copyPtr = 0;
+    for (int i = 0; i < numBuckets; i++) {
+      if (i == numBuckets - 1) {
+        // Special treatment for the last bucket
+        System.arraycopy(buckets.get(i), 0, compressedRecord, copyPtr, dataSize % bucketSize);
+      } else {
+        System.arraycopy(buckets.get(i), 0, compressedRecord, copyPtr, bucketSize);
+        copyPtr += bucketSize;
       }
     }
-    throw new HelixException(String.format("Could not acquire lock for read. Path: %s", path));
+
+    // Decompress the byte array
+    ByteArrayInputStream byteArrayInputStream = new ByteArrayInputStream(compressedRecord);
+    byte[] serializedRecord;
+    try {
+      serializedRecord = GZipCompressionUtil.uncompress(byteArrayInputStream);
+    } catch (IOException e) {
+      throw new HelixException(String.format("Failed to decompress path: %s!", path), e);
+    }
+
+    // Deserialize the record to retrieve the original
+    ZNRecord originalRecord = (ZNRecord) _zkSerializer.deserialize(serializedRecord);
+    return new HelixProperty(originalRecord);
+  }
+
+  @Override
+  public void close() {
+    disconnect();
   }
 
   /**
-   * Returns the last string element in a split String array by /.
+   * Deletes all versions that are older than the currentVersion given.
    * @param path
-   * @return
+   * @param currentVersion
    */
-  private String extractIdFromPath(String path) {
-    String[] splitPath = path.split("/");
-    return splitPath[splitPath.length - 1];
+  private void deleteStaleVersions(String path, long currentVersion) {
```

Comment:
I'm fine with the unbounded queue of the threadpool. But we don't really need this currentVersion, do we?

We can always read the most recent LAST_SUCCESSFUL_WRITE version before processing.

Reply:
I think the GC thread submitted immediately after the last successful update will likely introduce a *non-repeatable read* or *read skew* problem. 
Think about another transaction T2 firstly reads the data of last successful write version = 0 concurrently, T1 finishes the last step, try to update the last successful version = 1 and cleaned up the older versions (including 0),m committed. T2 wouldn't be able to read the version = 0 repeatedly after T1 commits.

Reply:
@i3wangyi That's why we have TTL for versions.

Reply:
@narendly For the 1) point, we can use batch read to minimize the extra latency. 2) point, I think it would preferred if the thread can delete a more recent version, as long as it is not the current one. There won't be a clean way to ensure all the old versions are cleaned up. There will be a race-condition. And the propagation latency of ZK might make it worse. So my point is to decouple the cleanup logic from the business as much as possible. So no matter when we want to trigger a cleanup, we can just call the method without worrying about side effects.

Reply:
Talked with @narendly offline, the idea of zookeeper TTL node interests me. I found starting from 3.5.3 zookeeper does support TTL nodes https://zookeeper.apache.org/doc/r3.5.4-beta/zookeeperProgrammers.html#TTL+Nodes. 

If using this feature, we can get rid of the custom GC thread at all. Here are some naive steps:
1. When a writer creates a node, it needs to be PERSISTENT
2. However, by checking the states of the nodes older than the newly created nodes. Update these nodes with TTL.

It sounds good in theory (the controller code looks clean, no GC thread), the concern is in reality, we don't even have that api supported. The code changes are even more. Just FYI.